File: ios/Classes/Dialer.swift
import Foundation
import Gomobile

class Dialer {
  typealias OnResult = (_ error: Error?) -> Void
  
  private let dialer:FormobileDialerProtocol = FormobileGetDialer()!
  let dialQueue = DispatchQueue(label: "go.dial")
  private var go:GomobileFromGoProtocol? = nil
  private var destroyRunner:GomobileConcurrentRunnerProtocol? = nil
  
  func doInit(_ onResult: @escaping OnResult) {
    if go == nil {
      go = GomobileNewFromGo()
    }
    
    let runner = go!.doInitOnce()!
    dialQueue.async {
      do {
        try runner.done()
        DispatchQueue.main.async { onResult(nil) }
      } catch {
        DispatchQueue.main.async { onResult(error) }
      }
    }
  }
  
  func doDestroy(_ onResult: @escaping OnResult) {
    guard let go = self.go else {
      onResult(nil)
      return
    }
    self.go = nil
    
    let runner = go.doDestroyOnce()!
    destroyRunner = runner
    dialQueue.async {
      do {
        try runner.done()
        DispatchQueue.main.async {
          onResult(nil)
          self.destroyRunner = nil
        }
      } catch {
        DispatchQueue.main.async {
          onResult(error)
          self.destroyRunner = nil
        }
      }
    }
  }
  
  func isNotInitialized() -> Bool {
    return go == nil
  }
  
  func dial(port: Int32, channelId: Int64, timeoutnano:Int64) throws -> FormobileConnProtocol {
    return try dialer.dial(port, channelId:channelId, timeoutnano: timeoutnano)
  }
}

File: android/src/main/java/io/github/empirefox/flutter_dial_go/Dialer.java
package io.github.empirefox.flutter_dial_go;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;

import formobile.Conn;
import formobile.Formobile;
import gomobile.ConcurrentRunner;
import gomobile.FromGo;
import gomobile.Gomobile;

public class Dialer extends Handler implements formobile.Dialer {
  private final formobile.Dialer dialer = Formobile.getDialer();
  private final HandlerThread dialThread;
  private final Handler dialHandler;
  private FromGo go = null;
  private ConcurrentRunner destroyRunner = null;

  public Dialer(Looper looper) {
    super(looper);
    dialThread = new HandlerThread("dial_thread");
    dialThread.start();
    dialHandler = new Handler(dialThread.getLooper());
  }

  public void doInit(Callback cb) {
    if (go == null) {
      go = Gomobile.newFromGo();
    }
    ConcurrentRunner runner = go.doInitOnce();
    dialHandler.post(() -> {
      try {
        runner.done();
        post(() -> cb.onResult(null));
      } catch (Exception e) {
        post(() -> cb.onResult(e));
      }
    });
  }

  public void doDestroy(Callback cb) {
    if (go == null) {
      cb.onResult(null);
      return;
    }

    FromGo go = this.go;
    this.go = null;
    destroyRunner = go.doDestroyOnce();

    dialHandler.post(() -> {
      try {
        destroyRunner.done();
        post(() -> {
          cb.onResult(null);
          destroyRunner = null;
        });
      } catch (Exception e) {
        post(() -> {
          cb.onResult(e);
          destroyRunner = null;
        });
      }
    });
  }

  public void doDestroySync() throws Exception {
    ConcurrentRunner runner = null;
    if (go != null) {
      runner = go.doDestroyOnce();
    }
    if (runner == null) {
      runner = destroyRunner;
      destroyRunner = null;
    }
    if (runner != null) {
      runner.done();
    }
    go = null;
  }

  public Handler getDialHandler() {
    return dialHandler;
  }

  public void stopDialThread() {
    dialThread.quit();
  }

  public boolean isNotInitialized() {
    return go == null;
  }

  @Override
  public Conn dial(int port, long channelId, long timeoutnano) throws Exception {
    return dialer.dial(port, channelId, timeoutnano);
  }

  public interface Callback {
    // fatal error if e is not null, bugs here!!!
    void onResult(Exception e);
  }
}
